### PR TITLE
Upgrade to fs2 1.0.0-M4

### PR DESCRIPTION
**Pull request by SystemFw**
Do not merge yet, we need an M4 because of https://github.com/functional-streams-for-scala/fs2/pull/1215

I've already made sure the tests pass with the fix though, so it should be good to go once M4 is out.

### CODE REVIEW

**Comment by SystemFw** [comment]
Heh, I was hoping to trigger the build again but it didn't work, can someone do it manually please?

**Comment by SystemFw** [comment]
Maybe it's my fault for always doing this late at night, but both relaxing the constraint to `ConcurrentEffect` and introducing `ContextShift` have caused test failures (it all works on my machine obviously, it's non deterministic).

If anything, `ContextShift` has made things worse (assuming that's the problem and not e.g. another regression in cats-effect or fs2).

In any case, we have a problem.

**Comment by SystemFw** [comment]
ah, I should mention that unlike the `take` bug, I haven't had time to dig into any of these failures in any depth

**Comment by SystemFw** [comment]
Ok, fixed everything and rebased the mess

File: core/src/main/scala/fs2/interop/reactivestreams/package.scala [hunk 1]
```diff
@@ -37,8 +34,22 @@ package object reactivestreams {
...
+    def toUnicastPublisher()(implicit F: ConcurrentEffect[F]): StreamUnicastPublisher[F, A] =
       StreamUnicastPublisher(stream)
   }
+
+  private[interop] implicit class Runner[F[_]: ConcurrentEffect, A](fa: F[A]) {
+    def reportFailure(e: Throwable) =
+      Thread.getDefaultUncaughtExceptionHandler match {
+        case null => e.printStackTrace()
+        case h => h.uncaughtException(Thread.currentThread(), e)
+      }
```
Comment (zainab-ali) [inline core/src/main/scala/fs2/interop/reactivestreams/package.scala:48]: This is a far, far better way of dealing with exceptions.  Thanks!

File: core/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala [hunk 2]
```diff
@@ -65,12 +66,7 @@ final class StreamSubscription[F[_], A](
   // See https://github.com/zainab-ali/fs2-reactive-streams/issues/29
   // and https://github.com/zainab-ali/fs2-reactive-streams/issues/46
   def cancel(): Unit =
-    IO.async[Unit] { cb =>
-        async.unsafeRunAsync {
-          cancelled.set(true) *> F.delay(cb(Right()))
-        }(_ => IO.unit)
-      }
-      .unsafeRunSync
+    cancelled.set(true).toIO.unsafeRunSync
```
Comment (zainab-ali) [inline core/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala:67]: :+1:

**Review by zainab-ali — APPROVED** [review]
Thanks so much for performing this upgrade!  Some of the additions to `cats-effect` have cleaned up the code nicely :smile: 

I hope you don't mind me pushing a few cosmetic changes to the branch (just a type and some scalafmt reformatting).

**Comment by SystemFw** [comment]
> I hope you don't mind me pushing a few cosmetic changes to the branch (just a type and some scalafmt reformatting).

Not at all :)

Unrelatedly, I hear you organise a meet up in London right? Just moved down here

**Comment by zainab-ali** [comment]
That I am!  It's been going for a few months now - we focus on functional open source contributions and learning in all sorts of languages.  The next meetup is focused on http4s.  Quite a few of us have been itching to contribute to it.  [You can sign up here](http://events.eventzilla.net/e/syntactic-sugar-functional-open-source-hacking-2138948799), if you'd like to come.  It would be great to see you there!

**Comment by rossabaker** [comment]
Thanks for fixing and merging.  This is a blocker to getting an http4s milestone out on the latest.  Will this be published anywhere?

Note that cats-effect-1.0 is imminent, so we'll probably be doing this again later this week.  Hopefully without the breaking changes. :smile:

Thanks!

**Pull request merged**